walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -63,9 +75,35 @@
 
     // Add available time after all evenets.
     if (start + meetingDuration <= TimeRange.END_OF_DAY) {
-      available.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, true));
+      optionalAvailable.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, true));
     }
+
+    // If there are times available with optional attendees or the meeting has no non-optional attendees, return.
+    if (!optionalAvailable.isEmpty() || meetingAttendees.isEmpty()) {
+      return optionalAvailable;
+    } else { 
+      // If there is no suitable time with optional attendees, try only with non-optional attendees.
     
-    return available;
+      // Sort events by start time and add available times to list.
+      Collections.sort(unavailable, TimeRange.ORDER_BY_START);
+      start = TimeRange.START_OF_DAY;
+      for (TimeRange time : unavailable) {
+        if (start + meetingDuration <= time.start()) {
+          available.add(TimeRange.fromStartEnd(start, time.start(), false));
+        }
+        // Check end time to account for when an event has an earlier start time but later end time.
+        if (time.end() > start) {
+          start = time.end();
+        }
+      }
+
+      // Add available time after all evenets.
+      if (start + meetingDuration <= TimeRange.END_OF_DAY) {
+        available.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, true));
+      }
```

Comment:
This looks like the exact same thing we did for the optionalUnavailable. Normally, if it's only repeated twice, I don't think pulling to a separate function is needed, but it also has enough complexity that pulling these out to a separate function would be helpful.

Reply:
Done.